rskj-core/src/main/java/co/rsk/net/sync/DownloadingBodiesSyncState.java

```diff
@@ -193,20 +196,20 @@ private void clearPeerInfo(NodeID peerId) {
 
     private Optional<BlockHeader> tryFindBlockHeader(NodeID peerId) {
         // we start from the last chunk that can be downloaded
-        for (int segmentNumber = segmentByNode.get(peerId);segmentNumber >= 0; segmentNumber--){
+        for (int segmentNumber = segmentByNode.get(peerId); segmentNumber >= 0; segmentNumber--){
             Deque<Integer> chunks = chunksBySegment.get(segmentNumber);
             // if the segment stack is empty then continue to next segment
             if (!chunks.isEmpty()) {
                 int chunkNumber = chunks.pollLast();
                 Deque<BlockHeader> headers = pendingHeaders.get(chunkNumber);
-                if (!headers.isEmpty()) {
-                    chunksBeingDownloaded.put(peerId, chunkNumber);
-                    segmentsBeingDownloaded.put(peerId, segmentNumber);
-                    return Optional.of(headers.poll());
-                } else {
-                    // log something we are in trouble
-                    syncEventsHandler.warnMessage("We got a segment {} with a unfinished chunk {} with no pending headers!", segmentNumber, chunkNumber);
-                    syncEventsHandler.stopSyncing();
+                while (!headers.isEmpty()) {
```

Comment:
FYI this is not thread safe

Reply:
Improved so if we need to change to a not thread safe execution it still works